### PR TITLE
Extended Static-Images Endpoint

### PR DESCRIPTION
**Pull request by benedikt-brandtner-bikemap**
I have Extended the Static-Images Endpoint with the following capabilities:

- Optional border with configurable color and width for paths via the `border` and `borderwidth` parameters
- Optional markers with the following configurable options:
  - Icon provided locally via the icons path or as a hyperlink (enabled via `allowRemoteMarkerIcons` setting)
  - Scale of the image relative to its original size
  - Offset in pixels
- Linecap of the path https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineCap
- Linejoin of the path https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineJoin
- Capability to provide multiple paths
- Maximum zoom level if autofit is selected when providing a path

Additional Changes:

- In order to keep performance drawing of elements has been refactored to be asynchronous
- Refactored a lot of helpers in serve_rendered.js to be more modular and reusable

Please let me know if this would be interesting to you.

Cheers :)

### CODE REVIEW

File: Dockerfile_test
Comment (acalcutt) [inline Dockerfile_test:27]: Does the docker need these new dependencies?

File: Dockerfile_test
Comment (mnutt) [inline Dockerfile_test:32]: Is there a reason we need to download this manually vs apt-getting it? Is it a versioning thing?

File: docs/endpoints.rst
Comment (mnutt) [inline docs/endpoints.rst:60]: I'm not opposed to this by any means, but have you considered just requiring that all markers be centered both horizontally and vertically on the coordinates? IMO it makes it really easy to get placement right when designing icons, and works really well with scaling, etc.

File: src/serve_rendered.js
Comment (mnutt) [inline src/serve_rendered.js:234]: I believe this would let someone specify a relative path that broke out of the icons dir and could pull from anywhere in the system. Maybe extract a function that checks that `iconURI.startsWith(options.paths.icons)`?

File: src/serve_rendered.js
Comment (mnutt) [inline src/serve_rendered.js:236]: We likely want to see if we can somehow avoid `existsSync` since it's in the request path. One way would be on startup to just synchronously pull a list of images in the icons dir, put them in an object, and retrieve their full paths here. It'd also solve the above issue around limiting the paths to the icons dir.

**Review by mnutt — COMMENTED** [review]
Thank you for this contribution! I'll be looking through it more thoroughly but wanted to leave a few initial comments.

File: src/serve_rendered.js
Comment (boldtrn) [reply src/serve_rendered.js:234]: I think we should definitely try to sanitize the input here. I would like to propose something like this: https://github.com/acalcutt/tileserver-gl/commit/64cad7da66eb1af01d079c547d69b2af990357d8

Not sure if we might need to be stricter and catch even more?

**Review by boldtrn — CHANGES_REQUESTED** [review]
Thank you very much for this great PR @benedikt-brandtner-bikemap. The features are really awesome 👍 .

I personally think that we should add a bit of security logic to this PR. Maybe we could make requesting external images configureable. For example if this is a public server, a malicious user might host a `https://` image somewhere and get the server to download this. At the very least it might be possible to force the server to download huge files and essentially DDOS the server. 

The same applies to the file name, in theory a malicious user could request some sensitive files and draw them on the map or in the worst case might even run some code injection.

WDYT?

**Comment by mnutt** [comment]
There's also the tricky concept of SSRF--allowing a server to take instruction to download any URL means it can fetch _internal_ resources that the server has access to (ec2 instances often have access to sensitive metadata on localhost, etc) and it's not very easy to set up rules to prevent it.

Realistically, siphoning internal resources through embedded markers is probably pretty difficult but would still potentially allow someone to prod around. I think it'd probably be best to leave the remote feature disabled by default, and let people enable it knowing the security implications.

**Comment by acalcutt** [comment]
Should this use a similar syntax to maptilers other products, like https://documentation.maptiler.com/hc/en-us/articles/360020805897-Static-maps-for-your-web . Seems like they support icon urls

File: Dockerfile_test
Comment (benedikt-brandtner-bikemap) [reply Dockerfile_test:32]: It seems like mablibre-gl depends on a version of libjpeg-turbo with legacy-capabilities enabled for version 8 which doesn't seem to be available in the current apt repository and also a libicu version which is not in the repository.

But for the dockerfile i just copied the packages from `Dockerfile`

File: docs/endpoints.rst
Comment (benedikt-brandtner-bikemap) [reply docs/endpoints.rst:60]: Maybe I don't understand the question, but the markers are already centered above the selected location https://github.com/maptiler/tileserver-gl/pull/619/files#diff-e18e0ca2c1587af2fb3550206da0889f4297c04bf4e3a1d1ee09ca46041a3c3dR51 and https://github.com/maptiler/tileserver-gl/pull/619/files#diff-51ea374a3960babfb87e15ca326fc54b011b08f981c33da61d174022970a5ce6R286
I guess it would be nice to have a `anchor` parameter similar to https://documentation.maptiler.com/hc/en-us/articles/360020805897-Static-maps-for-your-web#Staticmapsforyourweb-Addingmarkers but idk if that should be in the scope of this PR

**Comment by acalcutt** [comment]
Hope you are feeling better @benedikt-brandtner-bikemap . No rush here.

File: src/serve_rendered.js
Comment (benedikt-brandtner-bikemap) [reply src/serve_rendered.js]: have added the path sanitation with `sanitize-filename`

File: src/serve_rendered.js
Comment (benedikt-brandtner-bikemap) [reply src/serve_rendered.js]: all icons are now loaded into a settings object on server startup and the provided icon is checked against this list

**Comment by benedikt-brandtner-bikemap** [comment]
thanks a lot, but yeah am bored anyways so this is a welcome distraction :slightly_smiling_face: 

Have implemented the following fixes to address the mentioned issues:

1. Loading of remote images must now be allowed via the configuration parameter `allowRemoteMarkerIcons`
2. Locally available icons get loaded as relative paths to the icon-folder into the settings object `paths.availableIcons` on server startup (thanks a lot for the idea @mnutt)
3. When providing a local icon as a marker, the path gets sanitized first and then checked against the `paths.availableIcons` settings object if the icon exists and is allowed to be used

I think these measures should remedy a lot of the security concerns mentioned, although you are correct as soon as the icons get fetched from a remote server we are still exposed.

**Comment by benedikt-brandtner-bikemap** [comment]
Have rebased my changes ontop of current master to resolve conflicts in `Dockerfile_test`

File: src/server.js
Comment (boldtrn) [inline src/server.js:119]: Great idea 👍

**Comment by boldtrn** [comment]
> I think these measures should remedy a lot of the security concerns mentioned, although you are correct as soon as the icons get fetched from a remote server we are still exposed.

Thanks a lot for the changes. At least from my side, this solves the concerns I had and I am looking forward to using this 👍

**Comment by acalcutt** [comment]
I finally got a chance to set up a test of this. Seemed to work pretty well. it took me a bit to figure out the url.

I ended up with this for my test url
http://192.168.0.251:8080/styles/test-style/static/auto/800x600.png?&marker=5.9,45.9|icons8-maps.svg&marker=5.7,44.9|icons8-pin-62.png|scale:0.75&marker=5,45.1|icons8-pin-67.png|scale:0.75
and got
800x600

**Comment by acalcutt** [comment]
How do you use linecap?

**Comment by benedikt-brandtner-bikemap** [comment]
Hey, awesome that you approve of my changes if i should add/change anything else i'll be happy to do so!

@acalcutt I use linecap in the drawing context of the path if one is provided to define the rendering-style of the start- and end-points of the line as described [here](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineCap)
[implementation](https://github.com/maptiler/tileserver-gl/pull/619/files#diff-51ea374a3960babfb87e15ca326fc54b011b08f981c33da61d174022970a5ce6R414) 

Also you can find documentation for the new staticmap endpoint parameters (including markers) [here](https://github.com/maptiler/tileserver-gl/pull/619/files#diff-e18e0ca2c1587af2fb3550206da0889f4297c04bf4e3a1d1ee09ca46041a3c3d)

**Comment by benedikt-brandtner-bikemap** [comment]
Have updated documentation for the linecap parameter and also added a linejoin parameter as well to be able to alter the rendering style of overlapping segments as described [here](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineJoin)

**Comment by acalcutt** [comment]
Based on the docs and links, shouldn't this give me squared off line endings?

http://192.168.0.251:8080/styles/test-style/static/auto/800x600.png?&path=5.9,45.3|5.7,45.2&linecap=square
800x600 (3)

**Comment by benedikt-brandtner-bikemap** [comment]
Seems to work maybe you just cant see it because the width is too thin

default (butt):
http://127.0.0.1:8080/styles/bikemap-2021-3D/static/auto/800x600.png?width=10&path=16.370079517364502,48.2174529909167|16.3616681098938,48.21362115453811
default

square:
http://127.0.0.1:8080/styles/bikemap-2021-3D/static/auto/800x600.png?width=10&linecap=squ
are&path=16.370079517364502,48.2174529909167|16.3616681098938,48.21362115453811
square

round:
http://127.0.0.1:8080/styles/bikemap-2021-3D/static/auto/800x600.png?width=10&linecap=round&path=16.370079517364502,48.2174529909167|16.3616681098938,48.21362115453811
round


Also i've added the capability to draw multiple paths, which i wanted to do when i initially wrote the extensions but which came up again now.

sorry for extending the scope of this PR but it was only a minor change on top

**Comment by acalcutt** [comment]
Your right, it does work. i guess I expected more of a square at the ends. looks good.

**Comment by acalcutt** [comment]
@mnutt and @boldtrn  Any other things you would like to see before this is merged. it seems like most of the complaints have been fixed and it seems to work pretty well. I'd approve it and merge it in.

the only thing I wish we had was a default marker icon in our test style...but I think that would be something I would have to add to the old release zip file we use.

**Comment by boldtrn** [comment]
Default Marker would be nice indeed. So far I think this PR seems nice and I'd love to start using it.

**Comment by acalcutt** [comment]
I've been thinking a bit about the use case where you give tileserver a mbtiles file but do not give it a config. In that case it uses the built in style in '/node_modules/tileserver-gl-styles/', so it makes it somewhat hard to add the marker icons unless you add a config.

I was thinking, what if we put some default marker images in 'public/resources/icons/' and in main.js , add that icon path to the default generated config like this around line 125
```
      const styleDir = path.resolve(__dirname, '../node_modules/tileserver-gl-styles/');
      **const iconDir = path.resolve(__dirname, 'public/resources/icons/');**

      const config = {
        'options': {
          'paths': {
            'root': styleDir,
            'fonts': 'fonts',
            **'icons': iconDir,**
            'styles': 'styles',
            'mbtiles': path.dirname(mbtilesFile)
          }
        },
        'styles': {},
        'data': {}
      };
```

**Comment by acalcutt** [comment]
maybe something CC0 licensed for a default icon, like https://uxwing.com/map-pin-icon/

**Comment by acalcutt** [comment]
I was also thinking I wonder if the icons should be more like sprites so you could use a single image like https://github.com/sigma-geosistemas/Leaflet.awesome-markers/blob/master/dist/images/markers-matte.png with multiple colors in one file.

Not sure its needed to merge this, I was just wondering if that would be better. it seems like sprites are very similar to markers.

**Comment by mnutt** [comment]
I've looked over it and am good with it.

Regarding default markers: for what it's worth my old branch has a default marker that can be colorized based on a hex code:

https://github.com/mnutt/tileserver-gl/blob/master/src/markers/color.js

The 'template' uses solid red as the replacement color, and will dynamically generate PNGs of other colors.

**Comment by acalcutt** [comment]
> I've looked over it and am good with it.
> 
> Regarding default markers: for what it's worth my old branch has a default marker that can be colorized based on a hex code:
> 
> https://github.com/mnutt/tileserver-gl/blob/master/src/markers/color.js
> 
> The 'template' uses solid red as the replacement color, and will dynamically generate PNGs of other colors.

That is a really nice feature.

**Comment by benedikt-brandtner-bikemap** [comment]
Sorry for not beeing more active was focused on other tasks..

Is there anything I can do?

greets

**Comment by acalcutt** [comment]
What did you think of https://github.com/maptiler/tileserver-gl/pull/619#issuecomment-1279773329 , adding a default icon path.

**Review by acalcutt — APPROVED** [review]
Looks good to me. I still would like to see a default marker included, just so this can be used out of the box with the built in styles (like when just loading a mbtiles file with no config). then we could also use that default marker name in the documentation.

**Comment by benedikt-brandtner-bikemap** [comment]
Hey, sorry for not answering sooner was again occupied with something else...

Hmm I could implement a default marker with a choosable color utilizing @mnutt lib but I think I'll only have time end of next week so maybe we merge this PR and I'll open a new one with a default marker?

Greets

**Comment by acalcutt** [comment]
that sounds fine to me. I'll merge this.

**Comment by acalcutt** [comment]
Hey @benedikt-brandtner-bikemap ,

I noticed this after I merged, but would it be possible to make a PR that changes 

`import pkg from 'canvas'; `https://github.com/maptiler/tileserver-gl/blob/master/src/serve_rendered.js#L10
to
`import {createCanvas, Image} from 'canvas';`

and remove the line
`const {createCanvas, Image} = pkg;` https://github.com/maptiler/tileserver-gl/blob/master/src/serve_rendered.js#L25

**Comment by benedikt-brandtner-bikemap** [comment]
sorry just saw your comment, thanks for fixing this and merging :)!

**Pull request merged**
